src/cargo/ops/cargo_test.rs

```diff
@@ -129,11 +153,25 @@ fn build_and_run<'a>(manifest_path: &Path,
         try!(config.shell().verbose(|shell| {
             shell.status("Running", cmd.to_string())
         }));
-        match ExecEngine::exec(&mut ProcessEngine, cmd) {
-            Ok(()) => {}
-            Err(e) => return Ok(Err(e))
+
+        if let Err(e) = ExecEngine::exec(&mut ProcessEngine, cmd) {
+            errors.push(e);
+            if !options.no_fail_fast {
+                break
+            }
         }
     }
-
-    Ok(Ok(compile))
+    if errors.is_empty() {
+        Ok((compile, None))
+    } else if errors.len() == 1 {
+        // Just one error occured => we can return it.
+        Ok((compile, Some(errors.pop().unwrap())))
+    } else {
+        // Multiple tests failed => Create a more generic error
+        let err = process_error("Multiple tests failed",
+                                None,
+                                errors[0].exit.as_ref(),
+                                errors[0].output.as_ref());
```

Comment:
This may be better to use the `internal` error instead of specifically creating a new `process_error`